[FEATURE] Add seqan3::sam_flag enum and adapt alignment io.

smehringer opened the pull request:
Fixes #1345

h-2 commented on include/seqan3/io/alignment_file/misc.hpp:
```diff
@@ -21,4 +21,68 @@ namespace seqan3
+ */
+enum class sam_flag : uint16_t
+{
+   paired                  = 0x1,   //!< The aligned read is paired (paired-end sequencing).
```
Maybe add `none` with value zero so it can be initialised and reset without passing `{0}` which is not very expressive for an enum.

smehringer commented:
Since silly combinations are not formalized I would not do any checks by default. We might consider implementing a free function like `validate_sam_flags` that checks for common mistakes but this will not be part of this PR

rrahn commented on include/seqan3/io/alignment_file/format_bam.hpp:
```diff
@@ -775,7 +776,7 @@ inline void format_bam::write_alignment_record([[maybe_unused]] stream_type &  s
             /* bin         */ reg2bin(ref_offset.value_or(-1), ref_length),
             /* n_cigar_op  */ static_cast<uint16_t>(cigar_vector.size()),
-            /* flag        */ flag,
+            /* flag        */ static_cast<uint16_t>(flag),
```
the core should not hold a sam_flag type because the entire alignment record core is loaded as one binary dump?
Or is the field also read separately?

smehringer replied:
Yes, the entire alignment record core is loaded as one binary dump!

rrahn commented on include/seqan3/io/alignment_file/format_sam.hpp:
```diff
@@ -633,7 +651,7 @@ inline void format_sam::write_alignment_record(stream_type & stream,
                                                align_type && align,
                                                std::vector<cigar> const & cigar_vector,
-                                               uint16_t flag,
+                                               sam_flag flag,
```
is the flag being modified within the method? If not please mark it const.

rrahn commented on include/seqan3/io/alignment_file/format_sam.hpp:
```diff
@@ -633,7 +651,7 @@ inline void format_sam::write_alignment_record(stream_type & stream,
                                                std::vector<cigar> const & cigar_vector,
-                                               uint16_t flag,
+                                               sam_flag flag,
                                                uint8_t mapq,
```
same here if applicable.

smehringer replied:
I also changed this in format_bam

rrahn commented on test/unit/io/alignment_file/alignment_file_format_test_template.hpp:
```diff
@@ -15,7 +15,7 @@
 #include <seqan3/io/alignment_file/output.hpp>
 #include <seqan3/range/views/take.hpp>
-#include <seqan3/test/pretty_printing.hpp>
+// #include <seqan3/test/pretty_printing.hpp>
```
Some left over?

smehringer replied:
yes, thanks!

rrahn commented on test/snippet/io/alignment_file/sam_flags.cpp:
```diff
@@ -0,0 +1,38 @@
+#include <iostream>
+#include <sstream>
+
+#include <seqan3/core/type_list/type_list.hpp>
```
why do you need type_list?

smehringer replied:
Good catch :see_no_evil:

rrahn commented:
snippet tests are failing as well.

smehringer commented:
Note that I additionally added a debug_stream test for sam_flags since I added an overload.
I also updated the changelog. I need to clean the commit history before merging

rrahn commented on CHANGELOG.md:
```diff
@@ -58,6 +58,10 @@ If possible, provide tooling that performs the changes, e.g. a shell-script.
 
+#### I/O
+
+* The `field::FLAG` of SAM/BAM input and output is not an enum instead of a simple integer (see seqan3::sam_flag).
```
```suggestion
* The `field::FLAG` of SAM/BAM input and output is now an enum instead of a simple integer (see seqan3::sam_flag).
```
This is an API change. Can you please document how this will affect people and how to fix possible compiler errors.

smehringer replied:
I don't really know what to say here...? 
You cannot use integers anymore when writing but you need to static_cast your integers into the respective enum value?
And when you read, well you need to change all your types from integer to sam_flags wherever you use them..

smehringer replied:
@rrahn ?

rrahn commented on include/seqan3/io/alignment_file/format_bam.hpp:
```diff
@@ -394,7 +395,7 @@ inline void format_bam::read_alignment_record(stream_type & stream,
     }
 
-    flag = core.flag;                                                          // field::FLAG
+    flag = sam_flag{static_cast<uint16_t>(core.flag)};                         // field::FLAG
```
why is core.flag not already an uin16_t?

smehringer replied:
Because I copied the core structure from seqan2 aand I did not know that using bit fields is obsolete if you use the correct integers...

rrahn commented on test/snippet/io/alignment_file/sam_flags.cpp:
```diff
@@ -0,0 +1,37 @@
+    {
+        // Check if a certain flag value (bit) is set:
+        if (static_cast<bool>(seqan3::get<seqan3::field::FLAG>(rec) & seqan3::sam_flag::unmapped))
+        {
```
you could remove the braces

rrahn commented on test/unit/core/debug_stream_test.cpp:
```diff
@@ -260,3 +261,13 @@ TEST(debug_stream_test, named_enumeration)
+
+TEST(debug_stream_test, sam_flags)
+{
+    std::ostringstream o;
```
```suggestion
    std::ostringstream o{};
```

rrahn commented on include/seqan3/io/alignment_file/format_bam.hpp:
```diff
@@ -395,7 +395,7 @@ inline void format_bam::read_alignment_record(stream_type & stream,
     }
 
-    flag = sam_flag{static_cast<uint16_t>(core.flag)};                         // field::FLAG
+    flag = core.flag;                                                          // field::FLAG
```
👍 😄
